src/webob/acceptparse.py

```diff
@@ -255,11 +294,382 @@ class AcceptEncoding(Accept):
+                )
+            if isinstance(value, (tuple, list)):
+                result = []
+                for element in value:
+                    if isinstance(element, (tuple, list)):
+                        element = _item_qvalue_pair_to_header_element(
+                            pair=element
+                        )
+                    result.append(element)
+                header_str = ', '.join(result)
+            else:
+                header_str = str(value)
+        return header_str
+
+
+class AcceptLanguageValidHeader(AcceptLanguage):
+    """
+    Represent a valid ``Accept-Language`` header.
+
+    A valid header is one that conforms to :rfc:`RFC 7231, section 5.3.5
+    <7231#section-5.3.5>`.
+
+    We take the reference from the ``language-range`` syntax rule in :rfc:`RFC
+    7231, section 5.3.5 <7231#section-5.3.5>` to :rfc:`RFC 4647, section 2.1
+    <4647#section-2.1>` to mean that only basic language ranges (and not
+    extended language ranges) are expected in the ``Accept-Language`` header.
+
+    This object should not be modified. To add to the header, we can use the
+    addition operators (``+`` and ``+=``), which return a new object (see the
+    docstring for :meth:`AcceptLanguageValidHeader.__add__`).
+    """
+
+    # RFC 7231 Section 5.3.5 "Accept-Language":
+    # Accept-Language = 1#( language-range [ weight ] )
+    # language-range  =
+    #           <language-range, see [RFC4647], Section 2.1>
+    # RFC 4647 Section 2.1 "Basic Language Range":
+    # language-range   = (1*8ALPHA *("-" 1*8alphanum)) / "*"
+    # alphanum         = ALPHA / DIGIT
+    lang_range_re = r"""
+        \*|
+        (?:
+        [A-Za-z]{1,8}
+        (?:-[A-Za-z0-9]{1,8})*
+        )
+    """
+    lang_range_n_weight_re = _item_n_weight_re(item_re=lang_range_re)
+    lang_range_n_weight_compiled_re = re.compile(
+        lang_range_n_weight_re,
+        re.VERBOSE
+    )
+    accept_language_compiled_re = _list_1_or_more__compiled_re(
+        element_re=lang_range_n_weight_re,
+    )
+
+    def __init__(self, header_value):
+        """
+        Create an :class:`AcceptLanguageValidHeader` instance.
+
+        :param header_value: (``str``) header value.
+        :raises ValueError: if `header_value` is an invalid value for an
+                            ``Accept-Language`` header.
+        """
+        self._header_value = header_value
+        self._parsed = list(self.parse(header_value))
+        self._parsed_nonzero = [
+            (language_range, qvalue) for language_range, qvalue in self.parsed
+            if qvalue
+        ]
+
+    @property
+    def header_value(self):
+        """(``str``) The header value."""
+        return self._header_value
+
+    @property
+    def parsed(self):
+        """
+        (``list``) Parsed form of the header.
+
+        A list of (language range, quality value) tuples.
+        """
+        return self._parsed
+
+    @classmethod
+    def parse(cls, value):
```

Comment:
> And in this design, the subclasses are not really just implementation detail any more. For example, we can check whether a header is an invalid or no header with e.g. isinstance(header_object, AcceptLanguageInvalidHeader), and that is by design. I definitely started off wanting to hide everything behind the AcceptLanguage class, but it was getting too complex, and the functionality seems to be much clearer as objects of three distinct classes. (Right now the Accept-Language header in WebOb is two classes, AcceptLanguage and NilAccept, which are not even subclasses of the same base class, and have no connection with each other.)

Checking `isinstance` is considered to be poor form in most any subclassing hierarchy and is a hacky escape hatch. If you know up front that an `AcceptLanguage` header can be in one of these 3 states then the parent class should expose enough information to determine that without having to check the type of class you're getting. I really do think that these classes should be implementation details just like the old design of `AcceptLanguage` and `NilAccept`. The primary interface should be a function that parses strings (or other python objects like `python_object_to_header`) and returns *something* implementing the `AcceptLanguage` api. The fact that may return different classes is just an implementation convenience and the api design should not be dependent on it.